Merge highlevel3 into integration

xaterz opened the pull request:
What has been done in this branch:
* Implemented the 'Modules that are taken prior to other modules' page (shows all pairs)
* Implemented search function for 'number of students who took a specified module A prior to a specified module B in a target AY-Sem'

nlzz22 commented on components/model.py:
```diff
@@ -656,3 +656,96 @@ def get_all_mods_taken_together():
     DB_CURSOR.execute(sql_command)
```
please pylint this file

tgqiang commented on templates/modulesTakenPriorToOthers.html:
```diff
@@ -0,0 +1,100 @@
+		<p class="text-center">Shows all pairs of modules where module A is taken prior to module B</b></p>
+		<br>
+		<div class="modules-taken-prior-search-form">
+			<form action="/moduleTakenPriorToOthers" method="POST">
```
Validation is needed for both module code inputs.

You want to leave it to UI department to fix or you want to implement it?

tgqiang commented on templates/modulesTakenPriorToOthers.html:
```diff
@@ -0,0 +1,100 @@
+			</tbody>
+		</table>
+
+		<form style="margin-top: 25px;" action="/moduleTakenPriorToOthers" method="post">
```
Why does this <code>Go Back</code> navigation use the POST method? A GET method makes more sense.

tgqiang commented on templates/modulesTakenPriorToOthers.html:
```diff
@@ -0,0 +1,100 @@
+		<br>
+		<div class="modules-taken-prior-search-form">
+			<form action="/moduleTakenPriorToOthers" method="POST">
+				<p>Tell me the number of students who have taken</p>
```
"Tell me the number of students who have taken **A** before they are *going to* take **B** in **a specified AY-Semester**"

I tried injecting a <code>studentplan</code> row with the following: ['D5000000A', true, 'CS1020', 'AY 16/17 Sem 2']
(meaning student D5000000A has already taken CS1020 in AY 16/17 Sem 2)

This particular student was also taken into account for the query "Find students who have taken CS1010 before they are going to take CS1020 in AY 16/17 Sem 2".

Due to the **going to** clause, that student should not have been considered seeing that he has taken CS1020 in AY 16/17 Sem 2 **already**.

Thus, I wish to double-check whether students who HAVE TAKEN the module already should or should not be considered in this query.

tgqiang replied:
The implication of the above being either the UI changes, or the back-end changes.

xaterz commented:
Most of the issues raised have been fixed. Please review the commits from 4296b0e onwards.

My reply to some of the issues:

* I don't do test cases for UI. After this is approved, @tgqiang or @a0129998 please write the UI test cases for this feature, thanks.
* The reason why I always use POST method for navigation, is because if the GET method is used, there will be a trailing question mark in the URL, which doesn't look very nice.

![image](https://cloud.githubusercontent.com/assets/13976925/23825442/e2e6eb32-06c4-11e7-922b-0f60629c052f.png)

* But since this has been raised, I have changed the Go Back button to use javascript redirect instead. @tgqiang if you want to totally avoid using POST for navigation, you need to change for ALL the navigation buttons currently in our app.
* The 'going to take' issue is just a matter of phrasing. I have changed it to 'taking' instead. The ambiguity was actually explained in my code comment if you haven't notice:

![image](https://cloud.githubusercontent.com/assets/13976925/23825500/ec8c2c28-06c5-11e7-92df-e5683d4028af.png)

nlzz22 commented:
Pull request is approved: Issues raised previously have been addressed, all codes are linted and all test cases passed.
